tests/integration/utils2.go

```diff
@@ -824,24 +841,42 @@ func updateDoc(
 	var expectedErrorRaised bool
 	actionNodes := getNodes(action.NodeID, nodes)
 	for nodeID, collections := range getNodeCollections(action.NodeID, nodeCollections) {
-		// If a P2P-sync commit for the given document is already in progress this
-		// Save call can fail as the transaction will conflict. We dont want to worry
-		// about this in our tests so we just retry a few times until it works (or the
-		// retry limit is breached - important incase this is a different error)
-		for i := 0; i < actionNodes[nodeID].MaxTxnRetries(); i++ {
-			err = collections[action.CollectionID].Save(ctx, doc)
-			if err != nil && errors.Is(err, badgerds.ErrTxnConflict) {
-				time.Sleep(100 * time.Millisecond)
-				continue
-			}
-			break
-		}
+		err := withRetry(
+			ctx,
+			actionNodes,
+			nodeID,
+			func() error { return collections[action.CollectionID].Save(ctx, doc) },
+		)
 		expectedErrorRaised = AssertError(t, testCase.Description, err, action.ExpectedError)
 	}
 
 	assertExpectedErrorRaised(t, testCase.Description, action.ExpectedError, expectedErrorRaised)
 }
 
+// withRetry attempts to perform the given action, retrying up to a DB-defined
+// maximum attempt count if a transaction conflict error is returned.
+//
+// If a P2P-sync commit for the given document is already in progress this
+// Save call can fail as the transaction will conflict. We dont want to worry
+// about this in our tests so we just retry a few times until it works (or the
+// retry limit is breached - important incase this is a different error)
+func withRetry(
+	ctx context.Context,
+	nodes []*node.Node,
+	nodeID int,
```

Comment:
suggestion: You could use `node *node.Node` instead and avoid having to pass 2 params.

Reply:
I was 50-50 about that one, in my head at least a lot of the test utils code does stuff like that - preferring the refactoring out of finding the correct thing instead of saving on param count.  Will have a look in a bit and then see what I feel like in the moment unless you have a strong preference.

- [ ] maybe just pass in node

Reply:
Leaving as-is